### PR TITLE
5590 using dynamic gasprice

### PR DESCRIPTION
**Pull request by secmask**

### CODE REVIEW

File: data/storage/bolt.go [hunk 1]
```diff
@@ -1022,13 +1023,13 @@ func (bs *BoltStorage) RemoveStalePendingActivities(tx *bolt.Tx, stales []common
 	return nil
 }
 
-//PendingSetRate return pending set rate activity
-func (bs *BoltStorage) PendingSetRate(minedNonce uint64) (*common.ActivityRecord, uint64, error) {
+//PendingActivityForAction return pending set rate activity
```
Comment (halink0803) [inline data/storage/bolt.go]: Here the pending activity is not only for setrate but for deposit also.

File: data/fetcher/blockchain.go [hunk 2]
```diff
@@ -12,5 +14,5 @@ type Blockchain interface {
 	FetchRates(atBlock uint64, currentBlock uint64) (common.AllRateEntry, error)
 	TxStatus(tx ethereum.Hash) (string, uint64, error)
 	CurrentBlock() (uint64, error)
-	SetRateMinedNonce() (uint64, error)
+	blockchain.MinedNonePicker
```
Comment (halink0803) [inline data/fetcher/blockchain.go]: It should be "Nonce" instead of "None"?

File: core/reserve_core.go [hunk 3]
```diff
@@ -273,6 +259,70 @@ func (rc ReserveCore) Deposit(
...
+	address, supported := exchange.Address(token)
+	if !supported {
+		return nil, fmt.Errorf("exchange %s doesn't support token %s", exchange.ID(), token.ID)
+	}
+	found, err := rc.activityStorage.HasPendingDeposit(token, exchange)
+	if err != nil {
+		return nil, err
+	}
+	if found {
+		return nil, fmt.Errorf("there is a pending %s deposit to %s currently, please try again", token.ID, exchange.ID())
```
Comment (halink0803) [inline core/reserve_core.go]: We return when we see a pending deposit here. Then we cannot replace it later. The comment below seems not correct

File: core/reserve_core.go [hunk 4]
```diff
@@ -273,6 +259,70 @@ func (rc ReserveCore) Deposit(
...
+	}
+
+	if err = sanityCheckAmount(exchange, token, amount); err != nil {
+		return nil, err
+	}
+
+	// if there is a pending deposit tx, we replace it
```
Comment (halink0803) [inline core/reserve_core.go]: We cannot reach here if there is a pending deposit.

File: data/fetcher/blockchain.go (hunk 2 above)
Comment (secmask) [reply data/fetcher/blockchain.go]: oh, that's right

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -61,10 +56,11 @@ type Blockchain struct {
 	listedTokens []ethereum.Address
 	mu           sync.RWMutex
 
-	localSetRateNonce     uint64
-	setRateNonceTimestamp uint64
-	setting               Setting
-	l                     *zap.SugaredLogger
+	localSetRateNonce, localDepositNone          uint64
```
[the remark this answers is not in the document]
Comment (secmask) [reply blockchain/blockchain.go]: good catch

File: core/reserve_core.go (hunk 4 above)
Comment (secmask) [reply core/reserve_core.go]: I removed check pendingDeposit part

File: data/storage/bolt.go (hunk 1 above)
Comment (secmask) [reply data/storage/bolt.go]: fixed

File: data/fetcher/blockchain.go (hunk 2 above)
Comment (secmask) [reply data/fetcher/blockchain.go]: fixed

File: blockchain/blockchain.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (secmask) [reply blockchain/blockchain.go]: fixed

**Pull request merged**